openvpncloud/resource_user.go

```diff
@@ -0,0 +1,161 @@
+package openvpncloud
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/patoarvizu/terraform-provider-openvpn-cloud/client"
+)
+
+func resourceUser() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceUserCreate,
+		ReadContext:   resourceUserRead,
+		DeleteContext: resourceUserDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"username": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 120),
+			},
+			// Whatever value you set for role, it will be ignored during creation.
```

Comment:
But wouldn't it force recreate the resource if it detects a "change"? Or am I misunderstanding the issue?

Reply:
OpenVPN cloud acts in a weird way. You can set the `role` and `group_id` after creation and it won't force recreation even though the update is not supported.
At the moment, you can only Create, Read, and Delete.
https://openvpn.net/cloud-docs/api-guide/

By default, the API creates a user with `role=MEMBER` and `group_id=YOUR_DEFAULT_GROUP`.
```

  # openvpncloud_user.dog has changed
  ~ resource "openvpncloud_user" "test" {
      + group_id   = "403339e2-dbf4-4547-80f1-3bfd41eff5ce"
        id         = "test@test"
      + role       = "MEMBER"
        # (4 unchanged attributes hidden)
    }
```

Reply:
So if you have:

```
resource "openvpncloud_user" "test" {
  group_id = "123"
  id = "test@test"
  role = "MEMBER"
}
```

And run `apply`, the user will get created but `group_id` and `role` won't be set on the OpenVPN Cloud backend to what you set? And then if you run `apply` again, you'll see the output above? And then subsequent runs of `apply` will show no differences? Does it show `(known after apply)` when it first creates them?

What do you see if you run `terraform state show openvpncloud_user.test` between the first and the second apply? In other words, is it possible that these attributes are in the Terraform state after the first apply, so the second apply isn't strictly modifying them.

(I'm asking these questions because I'm at my job and can't verify this on my personal account.)

Reply:
Mmhh that's interesting. Two things:

1. I would consider the fact that `role` is ignored by the OpenVPN Cloud API as a bug on their side, but I don't know what are the guidelines/best practices for Terraform providers around bugs in upstream APIs.
2. I think the reason for the `id` changing is precisely because the unintended change in `role` is triggering a new resource (because the field is set with `ForceNew: true`), so the actual id on the OpenVPN Cloud backend will change because it'll create a new user. I bet if you keep running `terraform apply` over and over, without updating anything on the UI, it will keep re-creating the user, doesn't it?

I believe the right course of action here would be to implement a `resourceUserUpdate` function, except that it doesn't look like the API currently supports a `PUT` operation for the `/users` endpoint, at least according to the API guide or the Swagger interface. If you feel lucky, you could give it a try and see if maybe the API does support it but their Swagger file hasn't been updated, but it's unlikely.

So I think there are two options:

1. We just don't support setting/reading the `role` attribute via Terraform until they fix this issue on their backend (but of course you can still set it on the UI after the fact). We can print out a warning message regarding this, similar to the [warning message](https://github.com/patoarvizu/terraform-provider-openvpncloud/blob/main/openvpncloud/resource_connector.go#L75) we print for connectors.
2. We support the field but with the only possible value being `MEMBER` ([here](https://github.com/patoarvizu/terraform-provider-openvpncloud/blob/main/openvpncloud/resource_network.go#L41)'s an example of how to do that) and a warning that any change to the actual resource on the UI will be reverted to `MEMBER` on the next `terraform apply`.

I don't like the option of relying on a manual change after the initial creation to get things in sync. I personally think all Terraform code should set up things perfectly as declared, such that the very first `terraform apply` gets things created, and if I then immediately run `terraform apply` again, I should get zero differences. The options I outline above should be in line with that principle.

What do you think? I don't want to be prescriptive without getting your input first.

Reply:
>I bet if you keep running terraform apply over and over, without updating anything on the UI, it will keep re-creating the user, doesn't it?

Yes, it will. Setting `role` to `ADMIN` will force continuous recreation every time you run `terraform apply` unless you make changes manually via web UI. If you do it manually, terraform will check the role is set correctly and will not recreate the user.

>I personally think all Terraform code should set up things perfectly as declared, such that the very first terraform apply gets things created, and if I then immediately run terraform apply again, I should get zero differences.

I completely agree with your position on this.

>We support the field but with the only possible value being MEMBER ([here](https://github.com/patoarvizu/terraform-provider-openvpncloud/blob/main/openvpncloud/resource_network.go#L41)'s an example of how to do that) and a warning that any change to the actual resource on the UI will be reverted to MEMBER on the next terraform apply.

I like the idea of limiting the provider to creating users whose `role` is `MEMBER`. It will let us avoid post-apply changes in the resources. I will add this restriction today.

In the meantime, I will email the OpenVPN cloud support team and ask them about setting the role during the user's creation.